### PR TITLE
Lazy base types

### PR DESCRIPTION
**Pull request by JsonFreeman**
The spec section 3.9 does not say that interface / class types depend on their base types. And they shouldn't. However, the compiler eagerly resolves the base types of classes and interfaces, so if a class / interface is referenced by a type alias, it might lead to a circular reference error.

The issue can be seen in #1936.

The fix is to make collection of base types lazy, so that base type computation is done on demand when base types are needed.

### CODE REVIEW

**Comment by JsonFreeman** [comment]
I hope to make similar changes for indexers and return types of call and construct signatures.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2452,6 +2452,69 @@ module ts {
             return result;
         }
 
+        function getBaseTypes(type: InterfaceType): ObjectType[] {
+            if (!(<InterfaceTypeWithBaseTypes>type).baseTypes) {
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: we have ObjectType and ResolvedObjectType.  Should we have InterfaceType and ResolvedInterfaceType?

Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: you do the cast 4 times.  Can you just cast once to a local, and then use that from then on?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -2452,6 +2452,69 @@ module ts {
...
+
+            return (<InterfaceTypeWithBaseTypes>type).baseTypes;
+        }
+
+        function resolveBaseTypesOfClass(type: InterfaceTypeWithBaseTypes): void {
+            type.baseTypes = [];
+            let declaration = <ClassDeclaration>getDeclarationOfKind(type.symbol, SyntaxKind.ClassDeclaration);
+            let baseTypeNode = getClassExtendsHeritageClauseElement(declaration);
+            if (baseTypeNode) {
+                let baseType = getTypeFromHeritageClauseElement(baseTypeNode);
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: this can now be getTypeFromTypeNode to make things cleaner if you want.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -2452,6 +2452,69 @@ module ts {
...
+                    else {
+                        error(baseTypeNode, Diagnostics.A_class_may_only_extend_another_class);
+                    }
+                }
+            }
+        }
+
+        function resolveBaseTypesOfInterface(type: InterfaceTypeWithBaseTypes): void {
+            type.baseTypes = [];
+            forEach(type.symbol.declarations, declaration => {
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: consider for-of.  i prefer not to have unnecessary arrow function allocs.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -2452,6 +2452,69 @@ module ts {
...
+                if (declaration.kind === SyntaxKind.InterfaceDeclaration && getInterfaceBaseTypeNodes(<InterfaceDeclaration>declaration)) {
+                    forEach(getInterfaceBaseTypeNodes(<InterfaceDeclaration>declaration), node => {
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: same here.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -2452,6 +2452,69 @@ module ts {
...
+                        let baseType = getTypeFromHeritageClauseElement(node);
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: consider getTypeFromTypeNode.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -2647,9 +2671,10 @@ module ts {
             let constructSignatures = type.declaredConstructSignatures;
             let stringIndexType = type.declaredStringIndexType;
             let numberIndexType = type.declaredNumberIndexType;
-            if (type.baseTypes.length) {
+            let baseTypes = getBaseTypes(type);
+            if (baseTypes.length) {
                 members = createSymbolTable(type.declaredProperties);
-                forEach(type.baseTypes, baseType => {
+                forEach(baseTypes, baseType => {
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: consdier for-of.

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -10118,7 +10147,8 @@ module ts {
         }
 
         function checkInheritedPropertiesAreIdentical(type: InterfaceType, typeNode: Node): boolean {
-            if (!type.baseTypes.length || type.baseTypes.length === 1) {
+            let baseTypes = getBaseTypes(type);
+            if (!baseTypes.length || baseTypes.length === 1) {
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: baseTypes.length <= 1

**Comment by CyrusNajmabadi** [comment]
:+1:

File: src/compiler/checker.ts (hunk 7 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I was thinking the same thing when I saw this line

File: src/compiler/checker.ts (hunk 1 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I can use a local. The problem with ResolvedInterfaceType is that it implies that the type elements (members, call signatures, etc) are resolved, and that is not in fact the case.

File: src/compiler/checker.ts (hunk 3 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: yes

File: src/compiler/checker.ts (hunk 4 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: yes

File: src/compiler/checker.ts (hunk 6 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: yes

**Pull request merged**